### ISSUE TITLE
[BUG] Cutlass and Cute-DSL generate suboptimal code for UMMA that use more registers than necessary

### ISSUE THREAD

**Issue by tridao** [NVIDIA/cutlass#2408]
**Describe the bug**
The SASS code for UMMA uses more registers than necessary: the registers holding tmem descriptor and idesc keep changing between instructions. 
This is important as it affects Blackwell FMHA perf, e.g. causing a lot of register spill in Blackwell FMHA (example 77 and cute-dsl blackwell's fmha.py).

SASS code from example 71 (71_blackwell_gemm_with_collective_builder), we see the tmem (4th argument to UTCHMMA) and idesc registers keep changing between 4 instructions:
```
UTCHMMA gdesc[UR38], gdesc[UR40], tmem[UR17], tmem[UR26], idesc[UR27], UP2 ; 
UPLOP3.LUT UP2, UPT, UPT, UPT, UPT, 0x80, 0x8 ;   
UMOV UR35, 0x40004040 ;                                                            
UMOV UR33, 0x40004040 ;                                                            
UTCHMMA gdesc[UR34], gdesc[UR36], tmem[UR17], tmem[UR24], idesc[UR25], UPT ;       
UMOV UR31, 0x40004040 ;                                                            
UMOV UR29, 0x40004040 ;                                                            
UMOV UR43, 0x40004040 ;                                                            
UTCHMMA gdesc[UR30], gdesc[UR32], tmem[UR17], tmem[UR22], idesc[UR23], UPT ;      
UTCHMMA gdesc[UR42], gdesc[UR28], tmem[UR17], tmem[UR20], idesc[UR21], UPT ;      
```
SASS code from cute-dsl blackwell `dense_gemm_persistent.py`, similar:

<img width="1330" alt="Image" src="https://github.com/user-attachments/assets/9b9fb078-6c70-44f4-b050-446a7c5d15f0" />

**Steps/Code to reproduce bug**
Compile example 71 or cute-dsl blackwell dense_gemm_persistent.py and view the SASS

**Expected behavior**
The register holding tmem descriptor and idesc should stay the same between instructions.
This is the SASS from a gemm kernel in cublas (nvjet), where the tmem desc (4th argument to UTCHMMA) and idesc use the same registers throughout instructions:
<img width="1334" alt="Image" src="https://github.com/user-attachments/assets/f5392c03-439a-4c26-a7b5-cfc7d8d78d6f" />

**Environment details (please complete the following information):**
Newest Pytorch docker image: nvcr.io/nvidia/pytorch:25.05-py3
nvcc release 12.9, V12.9.41, Build cuda_12.9.r12.9/compiler.35813241_0

cc: @thakkarV

**Comment by fengxie** [comment]
Thanks for reporting and the very detailed explanation of the problem. We also discovered similar sub-optimal pattern when investigating DSL generated SASS in another case. The root cause might be similar.

We have a fix internally in DSL compiler to nvvm which improved the problem we looked. We will take a look at this pattern and see if more patches are required. We may also need to work with backend compiler team as well.

**Comment by tridao** [comment]
I think this is an issue with ptxas and likely not an issue with the DSL compiler. I've hand-written different versions of the UMMA ptx code and yet ptxas still generates this kind of SASS which uses lots of registers.
I've also looked at the SASS code from triton example (https://github.com/triton-lang/triton/blob/main/python/tutorials/gluon/01-attention-forward.py) with the same issue.
I've tried ptxas version 12.8.61, 12.8.93, 12.9.41, and the latest 12.9.86 (corresponding to CTK 12.9.1), all having the same issue.

**Comment by tridao** [comment]
If there's a fix, either with newer ptxas versions, or ptxas options, or a different way to write the ptx that generates good SASS, please lmk. I'm eager to try it, as this is important for Blackwell FMHA.

**Comment by fengxie** [comment]
Thanks a lot for experiments. If possible, do you mind sharing you hand-written PTX code for us to invest more? We can also share with PTXAS team for this pattern.

**Comment by thakkarV** [comment]
I have forwarded this to the compiler team, thanks a lot for the thorough bug report

**Comment by thakkarV** [comment]
@tridao have you tried previous versions of the CTK? I don't recall the SASS being deficient for prior versions. Wondering if you observe this change with any specific CTK version?

**Comment by tridao** [comment]
I've tried CTK 12.8.0, 12.8.1, 12.9.0, 12.9.1, all with the same issue. I'll try to write a short self-contained ptx file and post it there.

**Comment by fengxie** [comment]
> I've tried CTK 12.8.0, 12.8.1, 12.9.0, 12.9.1, all with the same issue. I'll try to write a short self-contained ptx file and post it there.

Thanks a lot for helping. We debugged a similar problem recently but not sure if it's the same problem. We were able to manually change PTX (also with IR compiler patch) to get efficient SASS pattern. We can't push the patch because just that patch only will make things even worse without fixing other issues.

**Comment by tridao** [comment]
Btw i've just tried w nvcc release 13.0, V13.0.48 and the issue is still there. Here's a snippet of the SASS after compiling example `77_blackwell_fmha_fp16`: 
```
UMOV UR63, 0x8200010 ;
UMOV UR37, 0x40004040 ;
UIADD3 UR4, UPT, UPT, UR32, 0x406, URZ ;
UTCHMMA gdesc[UR32], gdesc[UR36], tmem[UR47], tmem[UR62], idesc[UR63], !UPT ;
UMOV UR60, 0x0 ;
UMOV UR61, 0x8200010 ;
UMOV UR19, 0x40004040 ;
UIADD3 UR20, UPT, UPT, UR28, 0x28050, URZ ;
UMOV UR17, 0x40004040 ;
UMOV UR58, 0x0 ;
UMOV UR59, 0x8200010 ;
UMOV UR23, 0x40004040 ;
UMOV UR15, 0x40004040 ;
UMOV UR56, 0x0 ;
UMOV UR57, 0x8200010 ;
UMOV UR25, 0x40004040 ;
UMOV UR13, 0x40004040 ;
UTCHMMA gdesc[UR16], gdesc[UR18], tmem[UR46], tmem[UR60], idesc[UR61], UPT ;
```
We see e.g. the instruction descriptor is the same but get copied into UR63 and UR61.

**Comment by manishucsd** [comment]
What is this situation with this compiler issue? 

@hwu36 ?

**Comment by chriselrod** [comment]
FWIW, at Modular, we found that we could avoid this issue and trigger reuse of the idesc and tmem uniform registers via using inline PTX for the series of MMA instructions that should reuse these registers (e.g., one inline ptx string that contains 8 MMAs when the reduction dim is 128): https://github.com/modular/modular/blob/5257cb3b7cdecb7b3cb04c31d3d472c4e248211d/max/kernels/src/nn/mha_sm100_2q.mojo#L894-L938

**Comment by tridao** [comment]
> FWIW, at Modular, we found that we could avoid this issue and trigger reuse of the idesc and tmem uniform registers via using inline PTX for the series of MMA instructions that should reuse these registers (e.g., one inline ptx string that contains 8 MMAs when the reduction dim is 128): https://github.com/modular/modular/blob/5257cb3b7cdecb7b3cb04c31d3d472c4e248211d/max/kernels/src/nn/mha_sm100_2q.mojo#L894-L938

Ha that's what i've been doing as well. Will compare to your ptx, thanks!
https://github.com/Dao-AILab/flash-attention/blob/c8abdd432d3b020aad750f9f93f054cb438ec08a/flash_attn/cute/blackwell_helpers.py#L446

**Comment by manishucsd** [comment]
CTK 13.1 is out and produces better SASS for `UTCHMMA` regions. 

<img width="708" height="143" alt="Image" src="https://github.com/user-attachments/assets/b4145b8e-d052-454c-9fab-63bda6c53afd" />

**Comment by tridao** [comment]
Great, hopefully there's a version of cute-dsl that bundles ptxas 13.1 soon

**Comment by tridao** [comment]
Confirming that CTK 13.1 has fixed this issue

**Comment by manishucsd** [comment]
@tridao , I see that there is inline ptx for [the gemm](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/cute/blackwell_helpers.py#L95-L207) part in FA4? Does the fix actually work for you or you still bypass DSL cute.gemm call? 

On myside, I see that idesc is reused with CTK 13.1+, as it should be, but URs are still not used wisely by the compiler. A lot of `R2UR`, `MOV.SPILL`, `R2UR.FILL`

**Comment by tridao** [comment]
Yep i'm seeing the same. The compiler doesn't allocate URs very well and i'm doing a bunch of tricks but it's still not great.

**Comment by manishucsd** [comment]
I am annoyed and upset as an Ex-NVIDIAian and a kernel writer.

**Comment by manishucsd** [comment]
NVIDIA please re-open this issue and fix it. `UTCHMMA` is the most important Blackwell instruction. Let us have good codegen for it.

**Comment by vickiw973** [comment]
thanks for pointing this issue, we will double check on our side , hi @brandon-yujie-sun, can you help reopen this issue? thanks

**Comment by vickiw973** [comment]
Hi @manishucsd , just want to check, which example do you see the R2UR, MOV.SPILL, etc issue? thanks.

**Comment by manishucsd** [comment]
Hi @vickiw973 , As @tridao mentioned above this is in FA4. This should be there in most attention kernel implementations, where MMA warpgroup tries to donate registers to softmax. 

Some more points:
- The UTCHMMA path should use/reuse UR datapath without getting to R. Leave the Rs to be donated. The reason TMEM was added, i.e., MMA regions to be not register pressure heavy.
- One can control registers within inline ptx and `.reg` and that's what FA4 is doing, but for the UR path we really at the mercy of the compiler. 
- There is [shfl_sync](https://github.com/flashinfer-ai/flashinfer/blob/64ed071e23bf8d5d2d5af5c91577e5b8e036a1cf/csrc/trtllm_kernels/batched_gemm/trtllmGen_bmm_export/trtllm/dev/SmemTile.h#L29-L32) to trick compiler go down URF path, but it does not always work. I have tried it for the kernel where we see this issue. 

You can disasm FA4, below is a snippet:

```code
        /*2260*/                   UTCHMMA gdesc[UR6], gdesc[UR10], tmem[UR14], tmem[UR64], idesc[UR65], !UPT ;
        /*2270*/                   UMOV UR27, 0x40004040 ;
        /*2280*/                   UMOV UR9, UR55 ;
        /*2290*/                   UTCHMMA gdesc[UR4], gdesc[UR26], tmem[UR15], tmem[UR64], idesc[UR65], UPT ;
        /*22a0*/                   UMOV UR25, 0x40004040 ;
        /*22b0*/                   UMOV UR13, 0x8200490 ;
        /*22c0*/                   UTCHMMA gdesc[UR8], gdesc[UR24], tmem[UR16], tmem[UR64], idesc[UR65], UPT ;
        /*22d0*/                   UMOV UR21, 0x40004040 ;
        /*22e0*/                   UMOV UR19, 0x40004040 ;
        /*22f0*/                   UMOV UR17, 0x40004040 ;
        /*2300*/                   UIADD3 UR6, UPT, UPT, UR52, -0x800, URZ ;
        /*2310*/                   UMOV UR7, UR53 ;
        /*2320*/                   UIADD3 UR4, UPT, UPT, UR50, -0x800, URZ ;
        /*2330*/                   UMOV UR5, UR51 ;
        /*2340*/                   UIADD3 UR8, UPT, UPT, UR48, -0x800, URZ ;
        /*2350*/                   R2UR UR64, R16.reuse ;
        /*2360*/                   UIADD3 UR16, UPT, UPT, UR10, 0x402, URZ ;
        /*2370*/                   UTCHMMA gdesc[UR6], gdesc[UR20], tmem[UR14], tmem[UR12], idesc[UR13], UPT ;
        /*2380*/                   UMOV UR9, UR49 ;
        /*2390*/                   R2UR UR65, R16 ;
        /*23a0*/                   UTCHMMA gdesc[UR4], gdesc[UR18], tmem[UR15], tmem[UR12], idesc[UR13], UPT ;
        /*23b0*/                   UTCHMMA gdesc[UR8], gdesc[UR16], tmem[UR63], tmem[UR12], idesc[UR13], UPT ;
        /*23c0*/                   UIADD3 UR6, UPT, UPT, UR46, -0x800, URZ ;
        /*23d0*/                   UIADD3 UR14, UPT, UPT, UR10, 0x404, URZ ;
        /*23e0*/                   UMOV UR7, UR47 ;
        /*23f0*/                   UMOV UR15, 0x40004040 ;
        /*2400*/                   UIADD3 UR4, UPT, UPT, UR10, 0x406, URZ ;
        /*2410*/                   UIADD3 UR8, UPT, UPT, UR44, -0x800, URZ ;
        /*2420*/                   UMOV UR9, UR45 ;
        /*2430*/                   UTCHMMA gdesc[UR6], gdesc[UR14], tmem[UR64], tmem[UR12], idesc[UR13], UPT ;
        /*2440*/                   UMOV UR5, 0x40004040 ;
        /*2450*/                   R2UR.FILL UR12, R17 ;
        /*2460*/                   IMAD.U32 R17, RZ, RZ, UR22 ;
        /*2470*/                   UMOV UR6, 0x0 ;
        /*2480*/                   UMOV UR7, 0x8200490 ;
        /*2490*/                   R2UR.FILL UR13, R18 ;
        /*24a0*/                   UTCHMMA gdesc[UR8], gdesc[UR4], tmem[UR65], tmem[UR6], idesc[UR7], UPT ;
```

Expectation is close to nvjet GEMM UTCHMMA regions:

```code
00000006 28faf9b0	      UTCHMMA gdesc[UR16], gdesc[UR18], tmem[UR20], tmem[UR14], idesc[UR15], !UP1
00000006 28faf9c0	      UIADD3 UR16, UPT, UPT, UR16, 0x80, URZ
00000006 28faf9d0	      UIADD3 UR18, UPT, UPT, UR18, 0x2, URZ
00000006 28faf9e0	      UTCHMMA gdesc[UR16], gdesc[UR18], tmem[UR20], tmem[UR14], idesc[UR15], UPT
00000006 28faf9f0	      UIADD3 UR16, UPT, UPT, UR16, 0x80, URZ
00000006 28fafa00	      UIADD3 UR18, UPT, UPT, UR18, 0x2, URZ
00000006 28fafa10	      UTCHMMA gdesc[UR16], gdesc[UR18], tmem[UR20], tmem[UR14], idesc[UR15], UPT
00000006 28fafa20	      UIADD3 UR16, UPT, UPT, UR16, 0x80, URZ
00000006 28fafa30	      UIADD3 UR18, UPT, UPT, UR18, 0x2, URZ
00000006 28fafa40	      UTCHMMA gdesc[UR16], gdesc[UR18], tmem[UR20], tmem[UR14], idesc[UR15], UPT
```

**Comment by vickiw973** [comment]
thanks for the details, we will have a look.
